src/registry/cancellationtoken.cpp

```diff
@@ -0,0 +1,41 @@
+/****************************************************************************
+**
+** Copyright (C) 2015 Oleg Shparber
+** Copyright (C) 2013-2014 Jerzy Kozera
+** Contact: http://zealdocs.org/contact.html
+**
+** This file is part of Zeal.
+**
+** Zeal is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** Zeal is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with Zeal. If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "cancellationtoken.h"
+
+using namespace Zeal;
+
+CancellationToken::CancellationToken()
+{
+    m_cancelled = QSharedPointer<bool>(new bool(false));
+}
+
+void CancellationToken::cancel()
+{
+    *m_cancelled = true;
```

Comment:
There are no write races since mainwindow is the only place where the token gets cancelled.
And even if there would be many threads cancelling all set the value to true.
Finally, the shared pointer is not reassigned and cannot be disposed since at least one instance of CancellationToken exists (the one that is calling the cancel method).